2022-02-monthly-report.org

```diff
@@ -0,0 +1,466 @@
+
+In this example the macro has two rules so we demonstrate that we match the apropriate rule and transcribe it respectively.
+
+#+BEGIN_SRC rust
+macro_rules! add {
+    ($a:expr,$b:expr) => {
+        $a + $b
+    };
+    ($a:expr) => {
+        $a
+    };
+}
+
+fn main() -> i32 {
+    let mut x = add!(1);
+    x += add!(2, 3);
+
+    x - 6
+}
+#+END_SRC
+
+Another exmaple:
+
+#+BEGIN_SRC rust
+macro_rules! Test {
+    ($a:ident, $b:ty) => {
+        struct $a($b);
+    };
+}
+
+Test!(Foo, i32);
+
+fn main() -> i32 {
+    let a = Foo(123);
+    a.0 - 123
+}
+#+END_SRC
+
+Here we take into account the context of the macro invocation and parse it into AST::Items. In the even of failure to match a rule the compiler error looks like the following:
+
+#+BEGIN_SRC
+<source>:11:17: error: Failed to match any rule within macro
+    1 | macro_rules! add {
+      | ~                
+......
+   11 |     let mut x = add!(1, 2, 3);
+      |                 ^
+#+END_SRC
+
+More error handling has been added for when the transcribed rule actually is not fully used so for example:
+
+#+BEGIN_SRC
+<source>:4:9: error: tokens here and after are unparsed
+    4 |         struct BAD($b);
+      |         ^
+#+END_SRC
+
+see: https://godbolt.org/z/TK3qdG56n
+
+*** Range Lang items
+
+In rust ranges are turned into structs so what seems like piece of syntax to specify some kind of constraint is actually something which can be assigned and manipulated. This is one of the building blocks in our journey to support slices.
+
+#+BEGIN_SRC rust
+#[lang = "RangeFull"]
+pub struct RangeFull;
+
+#[lang = "Range"]
+pub struct Range<Idx> {
+    pub start: Idx,
+    pub end: Idx,
+}
+
+#[lang = "RangeFrom"]
+pub struct RangeFrom<Idx> {
+    pub start: Idx,
+}
+
+#[lang = "RangeTo"]
+pub struct RangeTo<Idx> {
+    pub end: Idx,
+}
+
+#[lang = "RangeInclusive"]
+pub struct RangeInclusive<Idx> {
+    pub start: Idx,
+    pub end: Idx,
+}
+
+fn test() {
+    let a = 1..2; // range
+    let b = 1..; // range from
+    let c = ..3; // range to
+    let d = 0..=2; // range inclusive
+}
+#+END_SRC
+
+See: https://doc.rust-lang.org/std/ops/struct.Range.html
+
+*** Index Lang items
+
+Another building block to support macros is the ability to suport the index lang item core::ops::index so that a range can be an argument and the code in core::slice::index can actually become the starting point in giving us a slice from an array.
```

Comment:
```suggestion
Another building block to support slices is the ability to suport the index lang item core::ops::index so that a range can be an argument and the code in core::slice::index can actually become the starting point in giving us a slice from an array.
```